ssm_jax/bp/gauss_bp.py

```diff
@@ -0,0 +1,212 @@
+from functools import partial
+from jax import jit
+from jax import numpy as jnp
+from jax.tree_util import tree_map
+
+
+def block_split(A, idx):
+    """Split square matrix A into four blocks at `idx`
+
+    For example splitting
+        [[0, 1, 2],
+         [3, 4, 5],
+         [6, 7 ,8]]
+    at idx = 2 produces:
+        ([[0,1],[3,4]],
+         [[2],[5]],
+         [[6,7]],
+         [[8]]).
+
+    Args:
+        A (D, D): array.
+        idx: int.
+    Returns:
+        A11 (dim1, dim1)
+        A12 (dim2, dim1)
+        A21 (dim1, dim2)
+        A22 (dim2, dim2)
+       where `dim1 + dim2 == D` and `dim2 = D - idx`.
+    """
+    split_array = jnp.array([idx])
+    vblocks = jnp.vsplit(A, split_array)
+    # [leaf for tree in forest for leaf in tree]
+    blocks = [block for vblock in vblocks for block in jnp.hsplit(vblock, split_array)]
+    # Can also do:
+    #   blocks = tree_map(lambda arr: jnp.hsplit(arr,split_array),vblocks)
+    # followed by a tree_flatten
+    return blocks
+
+
+def block_join(A11, A12, A21, A22):
+    return jnp.block([[A11, A12], [A21, A22]])
+
+
+def block_rev(A, idx):
+    blocks = block_split(A, idx)
+    return block_join(*blocks[::-1])
+
+
+def vec_swap(a, idx):
+    return jnp.concatenate((a[idx:], a[:idx]))
+
+
+def info_marginalise(K_blocks, hs):
+    """Calculate the parameters of marginalised MVN.
+
+    For x, y joint distributed as
+        p(x, y) = Nc(x,y| h, K),
+    the marginal distribution of x is given by:
+        p(y) = \int p(x, y) dx = Nc(y | hy_marg, Ky_marg)
+    where,
+        hy_marg = h2 - Kyx Kxx^{-1} h1
+        Ky_marg = Kyy - Kyx Kxx^{-1} Kxy
+
+    Args:
+        K_blocks: blocks of the joint precision matrix, (Kxx, Kxy, Kyy),
+                    Kxx (dim_x, dim_x),
+                    Kxy (dim_x, dim_y),
+                    Kyy (dim_y, dim_y).
+        hs (dim_x + dim_y, 1): joint precision weighted mean, (hx, hy):
+                    h1 (dim_x, 1),
+                    h2 (dim_y, 1).
+    Returns:
+        Ky_marg (dim_y, dim_y): marginal precision matrix.
+        hy_marg (dim_y,1): marginal precision weighted mean.
+    """
+    Kxx, Kxy, Kyy = K_blocks
+    hx, hy = hs
+    G = jnp.linalg.solve(Kxx, Kxy)
+    Ky_marg = Kyy - Kxy.T @ G
+    hy_marg = hy - G.T @ hx
+    return Ky_marg, hy_marg
+
+
+def info_condition(Kxx, Kxy, hx, y):
```

Comment:
I had avoided this in part because of vague recollections about jax not being very efficient at ignoring inputs when it could, but it seems like that might have been [resolved](https://github.com/google/jax/pull/6582).

It feels slightly strange to me to pass in an unused argument and I had envisioned this function as spending most of its time inside wrapper functions for conditioning the different types of potential objects. Having said that, very happy to follow your judgement on whatever you think would be most clear.

Reply:
Yes, it's a low level function, I guess we can leave it as is.